src/lib.rs

```diff
@@ -1,7 +1,15 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
-    }
+/// A marker trait representing types that have fixed layouts in memory
+///
+/// Types are `FixedLayout` if they have a well-defined layout in memory.
+/// This is usually reserved for primitive types like numbers and bool
+/// as well as complex type that are not `#[repr(rust)]` (the default)
+/// representation for Rust types
+pub unsafe trait FixedLayout {}
+pub fn ensure_fixed_layout<T: FixedLayout>() {}
+
+macro_rules! fixed_layout_impl {
+    ($($type:ty),*) => {
+        $(unsafe impl FixedLayout for $type {})*
+    };
 }
+fixed_layout_impl!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, bool);
```

Comment:
Here are some other types that seem like they should be included (maybe you're not for now on purpose?):
* `usize` and `isize` (or does `FixedLayout` require no platform differences?)
* `NonZero*`
* `Option<NonZero*>`

Reply:
Also, should `FixedLayout` apply to raw pointer types (`*const i32`)?

Reply:
Yes, raw pointers have a well defined layout, as do `references` and `Option<reference>`

Reply:
Yes I'm not done. Happy to accept PRs 😄

Reply:
Gotcha, that's alright, just wanted to check my understanding.

Reply:
@KrishnaSannasi I don't think that's entirely true. `&str` and `&[T]` aren't a fixed layout I believe...

Reply:
Right, I forgot about `!Sized`, references do have a fixed layout when the pointee is `Sized`

Reply:
Cool. I can add all these unless @AndrewGaspar wants to

Reply:
No, sorry, not right now, but maybe I'll contribute later. I'd like to use this crate as a public dependency of [another crate](https://github.com/rsmpi/rsmpi) once the API is semi-stable, so I'm just trying to keep an eye on the work here.